node/overseer/src/lib.rs

```diff
@@ -725,68 +763,171 @@ where
+				if initial_sync_finished {
+					// Initial sync is complete. Notify the subsystems and proceed to the main loop
+					let update = ActiveLeavesUpdate::start_work(ActivatedLeaf {
+						hash,
+						number,
+						status,
+						span,
+					});
+					self.broadcast_signal(OverseerSignal::ActiveLeaves(update)).await?;
+				}
 			}
 		}
 
-		loop {
+		// If initial sync is not complete - wait for it.
+		// This loop is identical to the main one but doesn't generate `ActiveLeaves` and `BlockFinalized` events until
+		// the initial full sync is complete.
+		// This is an infinite loop which executes only when `!initial_sync_finished`. The weird syntax is only to save
+		// one extra layer if indentation because it's already a bit tough for the eyes.
+		// Think about it like:
+		// ```
+		// if !initial_sync_finished {
+		//		loop {
+		//			select! {
+		//				...
+		//			}
+		// 		}
+		// }
+		//```
+		while !initial_sync_finished {
 			select! {
 				msg = self.events_rx.select_next_some() => {
 					match msg {
-						Event::MsgToSubsystem { msg, origin } => {
-							self.route_message(msg.into(), origin).await?;
-							self.metrics.on_message_relayed();
-						}
-						Event::Stop => {
-							self.stop().await;
-							return Ok(());
-						}
-						Event::BlockImported(block) => {
-							self.block_imported(block).await?;
-						}
-						Event::BlockFinalized(block) => {
-							self.block_finalized(block).await?;
-						}
-						Event::ExternalRequest(request) => {
-							self.handle_external_request(request);
-						}
+						Event::MsgToSubsystem { msg, origin } => self.handle_msg_to_subsystem(msg, origin).await?,
+						Event::Stop => return self.handle_stop().await,
+						Event::BlockImported(block) => _ = self.block_imported(block).await,
+						Event::BlockFinalized(block) if self.sync_oracle.finished_syncing() => {
+							// Send initial `ActiveLeaves`
+							let update = self.block_imported(block.clone()).await;
+							if !update.is_empty() {
+								self.broadcast_signal(OverseerSignal::ActiveLeaves(update)).await?;
+							} else {
+								// In theory we can receive `BlockImported` during initial major sync. In this case the
+								// update will be empty.
+								let span = match self.span_per_active_leaf.get(&block.hash) {
+									Some(span) => span.clone(),
+									None => {
+										// This should never happen.
+										gum::warn!(
+											target: LOG_TARGET,
+											?block.hash,
+											?block.number,
+											"Span for active leaf not found. This is not expected"
+										);
+										let span = Arc::new(jaeger::Span::new(block.hash, "leaf-activated"));
+										span
+									}
+								};
+								let update = ActiveLeavesUpdate::start_work(ActivatedLeaf {
+									hash: block.hash,
+									number: block.number,
+									status: LeafStatus::Fresh,
+									span,
+								});
+								self.broadcast_signal(OverseerSignal::ActiveLeaves(update)).await?;
```

Comment:
I don't get why we need this branch? When we are receving a block import notification while doing major sync, we should not even land in the match block here.

Reply:
If the finality has stalled and no `BlockImported` events are generated the subsystems in polkadot won't start because they initiate work on `ActiveLeavesUpdate`. 

Right now this is solved by issuing an `ActiveLeavesUpdate` on startup with the leaves from DB. This is causing problems though because subsystems start work for old leaves, the state from them is usually pruned and some subsystems generate a bunch of logs on stratup.

Then we decided to send this initial `ActiveLeavesUpdate` after the major sync is done. But here I see two cases:
1. Finality is stalled or we don't get `BlockImported` events for some reason. -> In this case we won't issue `ActiveLeavesUpdate` and the subsystems will be 'inactive'.
2. Everything is fine and we get `BlockImported` at some point after the initial major sync. We don't need any special handling in this case.

So the branch you are referring to tries to solve1. We just got `BlockFinalized`, the initial major sync is complete and I'm trying to guess if an `ActiveLeavesUpdate` will be generated.

What I do:
* Generate an artificial `BlockImported` event by calling `self.block_imported(..)`. It is supposed to generate non-empty `ActiveLeaves` in 99% of the cases but what if it doesn't?
* The fat else handles this 'what it doesn't' case with all the stuff that can go wrong.

I wrote tests to cover some of the cases and today I'm thinking about how to simplify this. I don't like it too :)

Reply:
> If the finality has stalled and no `BlockImported` events are generated the subsystems in polkadot won't start because they initiate work on `ActiveLeavesUpdate`.

I don't get the connection? When finality stalls, we will still import new blocks?

> 1. Finality is stalled or we don't get `BlockImported` events for some reason. -> In this case we won't issue `ActiveLeavesUpdate` and the subsystems will be 'inactive'.

This shouldn't happen.

I mean in general you could track when major sync is finished and then send the `ActiveLeavesUpdate` for all current leaves. However, I'm not really sure that we need this as there will always be some block import (yes with stalled finality it will be slower, but there will be blocks imported.

Reply:
> I don't get the connection? When finality stalls, we will still import new blocks?

For me 'finality has stalled' means 'no new blocks are generated'. I'm misunderstanding something here - can you explain why this is not true?

Reply:
We are never not producing any new blocks. This is not going to happen, we may slow down block production but not more. There is block production and then there is finality. The block production is for building new blocks and put them on top of the chain. Finality is the process of saying that a certain chain up to a certain point can not be modified anymore, because between the last finalized block and the tip of the chain you can have forks, re-orgs and also go back to older blocks to produce on them. However, the moment we have finalized a block, we can not go before this block.

Reply:
> We are never not producing any new blocks.

I see... so this means after the initial sync we'll always have `BlockImported` events and all the precautions in my PR make no sense?

Reply:
Maybe :eyes: 

I mean it can happen that we don't produce any blocks, but then we have much bigger problems as Parachain subsystems not working properly :grimacing:

Reply:
It's far more elegant this way!